openshift/WFLY-15421_OIDC_for_WildFly_s2i_v2.adoc

```diff
@@ -0,0 +1,306 @@
+
+* [ ] Managed domain
+
+* [x] OpenShift s2i
+
+* [ ] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+==== Galleon layer for oidc support
+
+This proposal relies on the WildFly `elytron-oidc-client` Galleon layer. This layer must be present when provisioning a server 
+in order for this feature to be available.
+
+==== Descriptor based configuration
+
+This support works out of the box and doesn’t require any addition to the WildFly s2i image. It applies to any kind of OIDC provider.
+
+Users can define expressions in the json file in order to inject values at runtime.
+For example, `provider-url: ${env.OIDC_URL}` would imply that the env variable `OIDC_URL=http://keycloak/auth/realms/WildFly` is set.
+
+NB: This way of configuring a deployment implies that a “client” has been created in the authorization server. That is expected.
+
+==== User CLI scripts called at packaging time
+
+When using the https://issues.redhat.com/browse/WFLY-14934[WildFly Maven plugin] to package an application, a script can be called to configure the OIDC subsystem for the deployed application.
+
+This support works out of the box and doesn’t require any addition to the WildFly s2i image. It applies to any kind of OIDC provider.
+
+This implies a set of WildFly CLI management operations to be called at packaging time.
+
+For example:
+`/subsystem=elytron-oidc-client/secure-deployment=simple-webapp.war:add(client-id=simple-webapp, public-client=true, provider-url=${OIDC_URL}, ssl-required=EXTERNAL)`
+
+NB: This way of configuring a deployment implies that a “client” has been created in the authorization server. That is expected.
+
+==== Environment variable based configuration
+
+This is the main topic of this proposal.
+
+A CLI script is generated at boot time to handle the OIDC subsystem configuration based on discovered 
+deployments (actually deployed and found in content repository or located in `standalone/deployments` directory).
+In addition there is an automatic creation and registration of the clients in the authorization server.
+
+In order for war deployments to be discovered they must have the `OIDC` auth-method in web.xml. 
+Deployments that contain a `WEB-INF/oidc.json` descriptor file are ignored.
+
+NB: A single provider can be configured for a running server. The discovered deployments are all registered inside the same OIDC provider.
+
+We are offering a new set of env variables `OIDC_*` to be used to configure the subsystem, enable the discovery of deployments and registration of 
+clients into the authorization server.
+
+The existing `SSO_*` env variables used for OIDC are usable for Keycloak OIDC integration. 
+It helps transition smoothly to use the native OIDC support. 
+This proposal defines a mapping between the `SSO_*` and `OIDC_*` env variables.
+The `SSO_*` env variables are deprecated.
+
+The env variable `OIDC_PROVIDER_NAME=<name>` MUST be set when using `OIDC_*` env variables. When using `SSO_*` env variables 
```

Comment:
Is this really needed? it is jut for identification which set of env variables to use. cannot be checked which variable are filled and use them? when I will use the RH-SSO then i should use also `keycloak` as name of provider? I think just SSO_PROVIDER should be enough and don't introduced OIDC variables.  When is filled SSO_PROVIDER then SSO_REAL and SSO_URL is not mandatory to use.

Reply:
The provider name is required to create and configure the elytron-oidc-client provider.
In the future we can imagine other kind of providers for which we would have to adjust some logic in the launch scripts to  implement the interaction with the provider (eg: add new client). The name would be used as a provider identifier.

Reply:
The provider name must be some exact name to configure elytron-oidc-client? for example `rh-sso` or `keycloak`. Will be supported also some non RH identity providers? Or the provider name cannot be null or empty to activate elytron-oidc-client configuration?

Reply:
@kstekovi , in the future, we can imagine to introduce new providers yes. But for now we are limiting this feature to keycloak/sso

Reply:
OpenID is a standard so it shouldn't matter which provider is used. cannot be used some general value like oidc?

`keycloak` is for SSO_XXX and keycloak adapter and `sso` is for OIDC_XXX and elytron-oidc-client?

Reply:
The env variable OIDC_XXX cover the standard, so yes these env variables will be used whatver the provider. Some logic (to interact with the management interface of the provider to create client) could be specific.
SSO_XXX are there to help migration from legacy Keycloak/SSO oidc adapter to elytron-oidc-client. OIDC_XXX env are used to configure the elytron-oidc-client subsystem whatever the oidc provider.

Reply:
Thank you. It make sense have both configuration divided into two sets as you write into AD.